### ISSUE TITLE
shortest_paths() terminates unexpectedly. 

### ISSUE THREAD

**Issue by kolumdium** [igraph/python-igraph#368]
**Describe the bug**
Unexpected crash of the python-kernel / python environment when calculating shortest_paths() for big enough graphs.

Under Linux the following error is thrown:
python3: ../../../source/igraph/src/dqueue.pmt:286: igraph_dqueue_push: Assertion `q->stor_begin != 0' failed.
Aborted

Under Windows it just dies silently.

Similar Issue to #28

**To reproduce**
```
import igraph as ig

def crash(n):
    g = ig.Graph()
    g.add_vertices(n)
    g.shortest_paths()
    print(n)
    
if __name__ == '__main__':
    for n in range(1, 10):
        print(n)
        crash(10 ** n)
```

**Version information**
I am using Version 0.8.3 of python-igraph.
I am running it in an anaconda Environment on Windows however I did get the packages with pip install.
Environment Details: [env.txt](https://github.com/igraph/python-igraph/files/5981700/env.txt)
(Windows 10 Home Insider Preview)
I also tried WSL Ubuntu 20.04 without anaconda, same Issue.

**Details**
I was trying to calculate the shortest paths in cities and was converting the city graphs to ig.Graph() when I noticed that some cities worked and some did not.
I boiled it down to the example provided above.
The python Kernel just dies when "n" gets big enough.
For me this was n = 5 so 10.000 entries.
I do understand that it is a lot of the shortest paths to calculate but I would expect some kind of warning or exception that the operation fails. Instead, it just dies silently.
I also tied it with the different subclasses of shortest paths calculations with the same result.

Is this behavior expected, reproducible or am I doing something wrong here?

Should I be able to use this package for graphs this size or should I look for alternatives?

I also looked at the Issues and it maybe related to 405, 1027, 1446, 1628 in addition to 28

**Comment by szhorvat** [comment]
Can you clarify what size graph it crashes with? 

I am not sure what you mean by "n = 10.000 entries".  The example you show goes up to 10^n with n=10, which is 10 billion (i.e. not reasonable).

**Comment by szhorvat** [comment]
I cannot reproduce this on macOS with 30,000 vertices. I don't have enough memory to test larger cases.

Note that with `k` vertices, you need `k^2*8` bytes _just to store the result_ (even forgetting the computation time, and memory for additional data structures). That's ~0.75 GB for 10,000 vertices, ~7 GB for 30,000 and 75 GB for 100,000. Thus one cannot go very high with all-pairs shortest paths.

**Comment by kolumdium** [comment]
Okey let me rephrase.

It crashes after n = 5 so it does work with 10000 verticies.
However n = 6 does not work.

I see that I missjudged the size.

My Issue with this is that the process dies silently (on Windows). 
Shouldn't it issue a warning if there is not enough memory?
Or atleast not crash the python-kernel?

**Comment by ntamas** [comment]
Crashing the Python kernel is definitely a problem; I'll look into what we can do here. The C core of igraph should return an `IGRAPH_ENOMEM` error code for out-of-memory conditions and it should simply be turned into an exception.

**Comment by ntamas** [comment]
@szhorvat has found a place in the codebase where `igraph_matrix_resize()` (which is used to allocate the required output matrix) does not check the return value of an underlying vector resize operation so an out-of-memory error can go undetected here. I'm pretty sure this is the reason why we are seeing an error here.

On the other hand, if you are planning to do shortest path calculations on graphs of this size, the best course of action is to call `shortest_paths()` for small batches of vertices only -- this way the matrix will be only of size N*M (N is the number of vertices, M is the batch size), and you can aggregate or postprocess the results in a later step. This would also allow parallelization over multiple Python processes.

**Comment by kolumdium** [comment]
Ok, thank you for your help.
I just want to spare someone else the search I underwent to find out what was wrong.

I am definetly going to rethink my approch though.

**Comment by ntamas** [comment]
The underlying issue has been fixed in `igraph` and `python-igraph` 0.9.0 already includes the fix so I'm closing this. Feel free to reopen if the issue resurfaces.
